### ISSUE TITLE
CSR should be an optional parameter

### ISSUE THREAD

**Issue by Shihta** [apache/trafficcontrol#3425]
CSR of API `deliveryservices/sslkeys/add` is a required parameter.
Refer to [here](https://traffic-control-cdn.readthedocs.io/en/latest/api/deliveryservices_sslkeys_add.html?highlight=csr) for details

However, it's not necessary.
It should be change to an optional parameter.

**Comment by JBevillC** [comment]
@Shihta Not sure if my opinion matters much in the Traffic Control community these days, but here is how I would handle it:

Structurally, there isn't much difference between a x509 Certificate and a x509 CertificateRequest.  The x509 CertificateRequest contains everything the x509 Certificate contains, with the exception of the any digital signatures, precomputed time stamps, and information about the Certificate Authority that signed it.

For this reason, here is my proposal for how the CSR should be handled. There are two cases: A) CSR field is empty B) CSR field is populated.  In both cases of A and B, the Certificate (CRT field) and Private Key (KEY field) are assumed to be valid and populated:
 
**Case A**: CSR field **is** empy in the input
1.  Certificate (CRT) and Private Key (KEY) match cryptographically and pass all other validation.
2.  Extract all Certificate Request ASN1 values from the Certificate, marshal this data into a CertificateRequest DER encoded binary array, PEM encode it, and then save it as the CSR field.  Essentially this operation is cloning the existing Certificate (CRT field) and discarding the fingerprint and any other information about the signing CA.  This operation would be the same as what the `openssl` tool can do with the `openssl x509 -x509toreq` function.

**Case B**: CSR field is **NOT** empty in the input
1.  Certificate Request (CSR) must be a valid pem block and parse into a x509.CertificateRequest() object.
2. CertificateRequest (CSR) Public Key values must match the Private Key (KEY field) public key values cryptographcally.
3. Certificate Request (CSR) Public Key values must also match those in the Certificate (CRT) as well. 
4. If steps (1), (2), and (3) succeed above, than allow the CSR to be saved to traffic vault.

Use Cases:
1. Provide the ability to extract the Certificate Request from a Certificate when CSR is empty.
2. Provide the ability to save custom/different Certificate Request that may have different attributes than the original Certificate (CRT), but validate that the key and certificate have the same public key values.

Thoughts?

**Comment by rawlinp** [comment]
@JBevillC I value your opinion a lot (especially when it comes to certificate stuff like this)! I think all the validation steps proposed for both cases make sense, but for Case A I'm thinking if the user doesn't provide a CSR to save through the API then they might not really care to have one generated from their provided cert. The sslkeys `/generate` endpoint provides that CSR generation workflow already, so it probably wouldn't be much extra code to generate a CSR when not given. From a Riak storage standpoint it might save a bit of storage in large-scale deployments that primarily get certs from LetsEncrypt (or something similar), but I think that storage is relatively small and wouldn't really be a problem.

So basically I agree with the validation for Case B, but for Case A I could go either way (either let the CSR field be empty or generate a CSR from the given cert).
